src/msgraph_core/tasks/large_file_upload.py

```diff
@@ -0,0 +1,273 @@
+
+    def upload_session_expired(
+        self, upload_session: Optional[LargeFileUploadSession] = None
+    ) -> bool:
+        now = datetime.now()
+        upload_session = upload_session or self.upload_session
+        if not hasattr(upload_session, "expiration_date_time"):
+            raise ValueError("Upload session does not have an expiration date time")
+        expiry = upload_session.expiration_date_time
+        if expiry is None:
+            raise ValueError("Expiry is None")
+        if isinstance(expiry, str):
+            then = datetime.strptime(expiry, "%Y-%m-%dT%H:%M:%S")
+        elif isinstance(expiry, datetime):
+            then = expiry
+        else:
+            raise ValueError("Expiry is not a string or datetime")
+        interval = now - then
+        if not isinstance(interval, timedelta):
+            raise ValueError("Interval is not a timedelta")
+        if interval.total_seconds() >= 0:
+            return True
+        return False
+
+    async def upload(self, after_chunk_upload: Optional[Callable] = None):
+        # Rewind at this point to take care of failures.
+        self.stream.seek(0)
+        if self.upload_session_expired(self.upload_session):
+            raise RuntimeError('The upload session is expired.')
+
+        self.on_chunk_upload_complete = after_chunk_upload or self.on_chunk_upload_complete
+        session = await self.next_chunk(
+            self.stream, 0, max(0, min(self.max_chunk_size - 1, self.file_size - 1))
+        )
+        process_next = session
+        # determine the range  to be uploaded
+        # even when resuming existing upload sessions.
+        range_parts = self.next_range[0].split("-") if self.next_range else ['0', '0']
+        end = min(int(range_parts[0]) + self.max_chunk_size - 1, self.file_size)
+        uploaded_range = [range_parts[0], end]
+        while self.chunks > 0:
+            session = process_next
+            try:
+                lfu_session: LargeFileUploadSession = session  # type: ignore
+                if lfu_session is None:
+                    continue
+                next_range = lfu_session.next_expected_ranges
+                old_url = self.get_validated_upload_url(self.upload_session)
+                lfu_session.upload_url = old_url
+                if self.on_chunk_upload_complete is not None:
+                    self.on_chunk_upload_complete(uploaded_range)
+                if not next_range:
+                    continue
+                range_parts = str(next_range[0]).split("-")
+                end = min(int(range_parts[0]) + self.max_chunk_size, self.file_size)
+                uploaded_range = [range_parts[0], end]
+                self.next_range = next_range[0] + "-"
+                process_next = await self.next_chunk(self.stream)
+            except Exception as error:
+                logging.error("Error uploading chunk  %s", error)
+            finally:
+                self.chunks -= 1
+        return session
```

Comment:
The upload of the last chunk will typically not be an upload session instance. You may end up returning an object that's not useful to the consumer. 

- https://github.com/microsoftgraph/msgraph-sdk-design/blob/master/tasks/FileUploadTask.md#large-file-upload-result-prototype
- https://learn.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0&preserve-view=true#completing-a-file